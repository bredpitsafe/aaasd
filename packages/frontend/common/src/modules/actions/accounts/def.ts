import type { TAccountCredentials, TRealAccount } from '../../../types/domain/account.ts';

type TUpdatableAccountCredentials = Omit<TAccountCredentials, 'id'> & {
    id?: TAccountCredentials['id'];
};

export enum EAccountCredentialsUpdateActions {
    Set = 'Set',
    KeepExisting = 'KeepExisting',
}

type TUpdatableAccountCredentialsSendBody = Omit<
    TUpdatableAccountCredentials,
    'secret' | 'passphrase'
> & {
    secret: TUpdatableAccountSecret;
    passphrase: TUpdatableAccountSecret;
};

export type TUpdatableRealAccountSendBody = Omit<TUpdatableRealAccount, 'credentials'> & {
    credentials: TUpdatableAccountCredentialsSendBody[];
};

export type TUpdatableAccountSecret =
    | { [EAccountCredentialsUpdateActions.Set]: string | null }
    | EAccountCredentialsUpdateActions.KeepExisting;

export type TUpdatableRealAccount = Partial<Omit<TRealAccount, 'id' | 'name' | 'credentials'>> & {
    id?: TAccountCredentials['id'];
    name: TRealAccount['name'];
    credentials: TUpdatableAccountCredentials[];
};

export const ACCOUNT_SECRET = '<secret>';
export const INTERNAL_NAME_PREFIX = 'internal.';
export const INTERNAL_KEY = 'InternalMarkets';
