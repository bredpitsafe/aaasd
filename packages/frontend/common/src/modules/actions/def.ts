import type { ISO, Milliseconds, Nanoseconds } from '@common/types';
import type { TraceId } from '@common/utils';

import type { TComponentId } from '../../types/domain/component.ts';
import type { TGate } from '../../types/domain/gates.ts';
import type { TRobot, TRobotId } from '../../types/domain/robots.ts';
import type { THandlerOptions } from '../communicationHandlers/def.ts';
import type { TUserName } from '../user';

export type TWithSnapshot = {
    // @deprecated
    isSnapshot: boolean;
    endOfSnapshot: boolean;
    startOfSnapshot: boolean;
};

export type TWithTraceId = {
    traceId: TraceId;
};

export type TUnsubscribe = {
    type: 'Unsubscribe';
};

export type TSubscribed = {
    type: 'Subscribed';
    platformTime: ISO | null;
};

export type SGate = Omit<TGate, 'type'>;

export enum EComponentCommands {
    StartComponent = 'StartComponent',
    StopComponent = 'StopComponent',
    UnblockRobot = 'UnblockRobot',
    DryRunComponent = 'DryRunComponent',
    GenericRobotCommand = 'GenericRobotCommand',
}

export type TComponentConfig = {
    type: 'ComponentConfig';
    user: TUserName;
    componentId: TComponentId;
    componentKind: string;
    componentName: string;
    platformTime: ISO;
    config: string;
    digest: string;
};

export type TRobotDashboard = {
    id: number;
    platformTime: ISO;
    robotId: TRobotId;
    robotName: string;
    name: string;
    backtestingId?: number;
    focusTo?: ISO;
    raw?: string;
};

export type TUnsubscribeSendBody = {
    type: 'Unsubscribe';
};

export type THandlerStreamOptions = THandlerOptions & {
    updatesOnly?: boolean;
    pollInterval?: Milliseconds;
};

export type TWithUpdatesOnly = {
    updatesOnly?: boolean;
};

export type TWithPollInterval = {
    pollInterval?: Milliseconds;
};

export type TWithServerPollInterval = {
    pollInterval?: Nanoseconds;
};

export type TRequestStreamOptions = Partial<{
    updatesOnly: boolean;
    /**
     * @description Only specific types are valid:
     * Duration::from_millis(50),
     * Duration::from_millis(100),
     * Duration::from_millis(200),
     * Duration::from_millis(500),
     * Duration::from_secs(1),
     * Duration::from_secs(2),
     * Duration::from_secs(5),
     * https://bhft-company.atlassian.net/browse/PLT-3015
     */
    pollInterval: Nanoseconds;
}>;

export enum EFetchHistoryDirection {
    Forward = 'Forward',
    Backward = 'Backward',
}

export type TFetchHistoryParams = {
    limit: number;
    direction: EFetchHistoryDirection;
    timestamp: ISO;
    timestampExcluded?: boolean;
    timestampBound?: ISO;
    timestampBoundExcluded?: boolean;
};

export type TServerFetchHistoryParams = {
    limit?: undefined | number;
    // Send softLimit items count + all with same time
    softLimit?: undefined | number;
    direction: EFetchHistoryDirection;
    platformTime: ISO;
    platformTimeExcluded?: boolean; // optional, false by default, excludes start time (platform_time) from response
    platformTimeBound?: ISO;
    platformTimeBoundExcluded?: boolean; // optional, false by default, excludes end time (platform_time_bound) from response if time bound reached
};

export enum ETimeseriesAggregationFunction {
    FlattenCandle = 1,
    Last = 2,
    Candle = 3,
    Unspecified = 0,
}

export enum ETimeseriesAggregationIntervalUnit {
    Unspecified = 0,
    Millis = 1,
    Sec = 2,
    Min = 3,
    Hour = 4,
}

export type TTimeseriesAggregation<
    F extends ETimeseriesAggregationFunction = ETimeseriesAggregationFunction,
> = {
    function: F;
    intervalUnit: ETimeseriesAggregationIntervalUnit;
    intervalValue: number; // uint32
};

export type TTimeseriesFilter = {
    // `bt_run_no` which will be used as default value. Will be override by bt_run_no from
    // PromQL query. If bt_run_no didn't provide in `default_bt_run_no` and PromQL - data
    // will be returned not from backtest.
    defaultBtRunNo: number;
    // **REQUIRED** PromQL request for timeseries. Contains to fields:
    // - bt_run_no: number of backtest run.
    // - name - regex expression to specify names of timeseries.
    // Example:
    //  - indicator{bt_run_no=42,name~="akrousdt\|HuobiSpot.l1.ask|akrousdt\|HuobiSpot.volume"}
    //  - indicator{name="arusdt|HuobiSpot.l1.bid"}
    stringQuery: string;
};

export type TFetchSnapshotParams = {
    limit: number;
    offset: number;
    withTotal?: boolean;
};

export enum EFetchSortOrder {
    Asc = 'Asc',
    Desc = 'Desc',
}

export type TFetchSortFieldOrder<T extends Record<string, unknown>> = [keyof T, EFetchSortOrder];
export type TFetchSortFieldsOrder<T extends Record<string, unknown>> = TFetchSortFieldOrder<T>[];

export const UPDATES_ONLY: Pick<THandlerStreamOptions, 'updatesOnly'> = {
    updatesOnly: true,
};

export const NO_RETRIES: Pick<THandlerOptions, 'enableRetries'> = {
    enableRetries: false,
};

export const NO_LOGS: Pick<THandlerOptions, 'enableLogs'> = {
    enableLogs: false,
};

export type TRobotConfig = Pick<TRobot, 'id'> &
    Partial<Pick<TComponentConfig, 'config' | 'digest'>>;

export type TRobotConfigRecord = Record<TRobotId, TRobotConfig>;
export const DEFAULT_RETRY_DELAY = 5_000 as Milliseconds;
