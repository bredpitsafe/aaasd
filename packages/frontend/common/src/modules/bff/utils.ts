import type { TModuleConstructor } from '@frontend/common/src/di';
import { ModuleBFF } from '@frontend/common/src/modules/bff';
import type { TWithSocketTarget } from '@frontend/common/src/types/domain/sockets.ts';
import type {
    TObservableProcedure,
    TObservableProcedureOptions,
    TObservableProcedureSettings,
} from '@frontend/common/src/utils/LPC/createObservableProcedure.ts';
import { createObservableProcedure } from '@frontend/common/src/utils/LPC/createObservableProcedure.ts';
import { switchMap } from 'rxjs';

export function createModuleSubscriptionWithCurrentBFFStage<P extends TWithSocketTarget, R>(
    SubscribeFactory: TModuleConstructor<TObservableProcedure<P, R>>,
    settings?: TObservableProcedureSettings<Omit<P, 'target'>>,
) {
    return createObservableProcedure((ctx) => {
        const { getCurrentBFFSocket$ } = ModuleBFF(ctx);
        const subscribe = SubscribeFactory(ctx);

        return (params: Omit<P, 'target'>, options: TObservableProcedureOptions) => {
            return getCurrentBFFSocket$().pipe(
                switchMap((target) => {
                    return subscribe({ ...params, target } as P, options);
                }),
            );
        };
    }, settings);
}
