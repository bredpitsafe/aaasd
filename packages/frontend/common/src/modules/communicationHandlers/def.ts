import type { Assign } from '@common/types';
import type { TraceId } from '@common/utils';
import type { Observable } from 'rxjs';

import type { TReceivedData, TSendPayload } from '../../lib/BFFSocket/def';
import type { TSocketStreamOptions } from '../../lib/SocketStream/def';
import type { TSocketStruct, TSocketURL } from '../../types/domain/sockets';
import type { TStructurallyCloneableObject } from '../../types/serialization';
import type { TValueDescriptor2 } from '../../utils/ValueDescriptor/types';
import type { TUnsubscribe, TWithTraceId } from '../actions/def.ts';

export type THandlerOptions = Omit<TSocketStreamOptions, 'traceId'> &
    Partial<TWithTraceId> & {
        timeout?: number;
        enableRetries?: boolean;
        retries?: number;
        retryOnReconnect?: boolean;
        enableLogs?: boolean;
        skipAuthentication?: boolean;
    };

export type TStreamSender<SendBody> = () => [SendBody] | [SendBody, TUnsubscribe];

export type TStreamHandler = <
    SendBody extends TSendPayload,
    ReceiveBody extends TStructurallyCloneableObject,
    EnableVD extends boolean = false,
>(
    target: TSocketURL | TSocketStruct,
    sender: TStreamSender<SendBody>,
    options?: THandlerOptions & { enableVD?: EnableVD },
) => Observable<
    EnableVD extends true
        ? TValueDescriptor2<TReceivedData<ReceiveBody>>
        : TReceivedData<ReceiveBody>
>;

export type TFetchHandler = <
    SendBody extends TSendPayload,
    ReceiveBody extends TStructurallyCloneableObject,
    EnableVD extends boolean = false,
>(
    target: TSocketURL | TSocketStruct,
    body: SendBody,
    options?: THandlerOptions & { enableVD?: EnableVD },
) => Observable<
    EnableVD extends true
        ? TValueDescriptor2<TReceivedData<ReceiveBody>>
        : TReceivedData<ReceiveBody>
>;

export type TRequiredHandlerOptions = Assign<
    THandlerOptions,
    typeof DEFAULT_OPTIONS & {
        traceId: TraceId;
    }
>;

export const DEFAULT_OPTIONS: Required<
    Pick<
        THandlerOptions,
        | 'enableRetries'
        | 'retries'
        | 'retryOnReconnect'
        | 'timeout'
        | 'enableLogs'
        | 'waitForResponse'
    >
> = {
    timeout: 10_000,
    enableRetries: true,
    retries: 3,
    retryOnReconnect: true,
    enableLogs: true,
    waitForResponse: true,
};
