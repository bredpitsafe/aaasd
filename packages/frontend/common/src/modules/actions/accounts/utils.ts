import type {
    TUpdatableAccountSecret,
    TUpdatableRealAccount,
    TUpdatableRealAccountSendBody,
} from './def.ts';
import { ACCOUNT_SECRET, EAccountCredentialsUpdateActions } from './def.ts';

export function getSendBodyAccounts(
    accounts: TUpdatableRealAccount[],
): TUpdatableRealAccountSendBody[] {
    return accounts.map(({ credentials, ...acc }) => ({
        ...acc,
        credentials: credentials.map((cred) => {
            return {
                ...cred,
                secret: getUpdatableAccountSecret(cred.secret),
                passphrase: getUpdatableAccountSecret(cred.passphrase),
            };
        }),
    }));
}
function getUpdatableAccountSecret(value: string | null | undefined): TUpdatableAccountSecret {
    if (value === ACCOUNT_SECRET) {
        return EAccountCredentialsUpdateActions.KeepExisting;
    }
    return { [EAccountCredentialsUpdateActions.Set]: value ? value : null };
}
