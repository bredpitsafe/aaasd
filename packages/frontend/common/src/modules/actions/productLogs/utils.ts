import { isNil } from 'lodash-es';

import { shallowHash } from '../../../utils/shallowHash';
import { logger } from '../../../utils/Tracing';
import type { TProductLog } from './defs.ts';

export function tryFixFingerprints(items: TProductLog[]): TProductLog[] {
    const existed = new Set<TProductLog['fingerprint']>();
    let everyHasFingerprint = true;

    for (let i = 0; i < items.length; i++) {
        const item = items[i];

        if (!isNil(item.dedupKey) && !isNil(item.dedupCount)) {
            // As dedupKey + dedupCount is not uniq calc fingerprint with several fields
            item.fingerprint = String(
                shallowHash(
                    item.dedupKey,
                    item.dedupCount,
                    item.actorGroup,
                    item.actorKey,
                    item.message,
                    item.level,
                    item.component,
                    item.platformTime,
                ),
            );
        }

        const hasOriginalFingerprint = !isNil(item.fingerprint);

        if (!hasOriginalFingerprint) {
            everyHasFingerprint = false;
            item.fingerprint = 'UI_' + String(shallowHash(JSON.stringify(item)));
        }

        if (existed.has(item.fingerprint)) {
            item.fingerprint = item.fingerprint + '_' + Math.random();
            hasOriginalFingerprint &&
                logger.error(`[ProductLogs]: duplicate with fingerprint ${item.fingerprint}`);
        }

        existed.add(item.fingerprint);
    }

    if (!everyHasFingerprint) {
        logger.error(`[ProductLogs]: empty fingerprints`);
    }

    return items;
}
