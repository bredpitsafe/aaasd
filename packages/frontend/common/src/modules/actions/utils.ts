import type { TSubscriptionEvent } from '@common/rx';
import { createSubscribedEvent, createUpdateEvent } from '@common/rx';
import type { Milliseconds, Nanoseconds, Nil, TimeZone } from '@common/types';
import { TimeZoneList } from '@common/types';
import type { TraceId } from '@common/utils';
import { generateTraceId, isISO, milliseconds2nanoseconds } from '@common/utils';
import { isNil } from 'lodash-es';
import type { OperatorFunction } from 'rxjs';
import { filter } from 'rxjs/operators';

import type { TReceivedData } from '../../lib/BFFSocket/def.ts';
import type { TBacktestingRunId } from '../../types/domain/backtestings.ts';
import type { TGate, TGateType } from '../../types/domain/gates.ts';
import { mapValueDescriptor } from '../../utils/Rx/ValueDescriptor2.ts';
import type { TValueDescriptor2 } from '../../utils/ValueDescriptor/types.ts';
import {
    createSyncedValueDescriptor,
    isSyncedValueDescriptor,
} from '../../utils/ValueDescriptor/utils.ts';
import type { SGate, TSubscribed } from './def.ts';
import type { TIndicator, TIndicatorKey } from './indicators/defs.ts';

export function serverGateToGate(gate: SGate, type: TGateType): TGate {
    return { ...gate, type };
}

export function pollIntervalForRequest(pollInterval?: Milliseconds): Nanoseconds | undefined {
    if (isNil(pollInterval)) {
        return undefined;
    }

    return milliseconds2nanoseconds(pollInterval);
}

export function getTraceId<T extends { traceId?: TraceId }>(obj?: undefined | T): TraceId {
    return obj?.traceId ?? generateTraceId();
}

export function getIndicatorKey(
    url: TIndicator['url'],
    name: TIndicator['name'],
    backtestingRunId: Nil | TBacktestingRunId,
): TIndicatorKey {
    return `{${url}}.{${name}}.{${backtestingRunId || 0}}` as TIndicatorKey;
}

export function filterOutSubscribedValueDescriptor<T extends { type: string }>(): OperatorFunction<
    TValueDescriptor2<TReceivedData<T | TSubscribed>>,
    TValueDescriptor2<TReceivedData<T>>
> {
    return filter(
        (
            response: TValueDescriptor2<TReceivedData<T | TSubscribed>>,
        ): response is TValueDescriptor2<TReceivedData<T>> => {
            return (
                isSyncedValueDescriptor(response) && response.value.payload.type !== 'Subscribed'
            );
        },
    );
}

export function timeZone2UtcOrMskTimeZone(timeZone: TimeZone): string {
    switch (timeZone) {
        case TimeZoneList.UTC:
            return 'Utc';
        case TimeZoneList.EuropeMoscow:
            return 'Msk';
        default:
            throw new Error(`Time zone "${timeZone}" is not supported on backend`);
    }
}

export function convertToSubscriptionEventValueDescriptor<P extends object, R>(
    getResult: (payload: Exclude<P, TSubscribed>) => R,
) {
    let subIndex = 0;

    return mapValueDescriptor<
        TValueDescriptor2<TReceivedData<P>>,
        TValueDescriptor2<TSubscriptionEvent<R>>
    >(({ value: envelope }) => {
        const value =
            'type' in envelope.payload && envelope.payload.type === 'Subscribed'
                ? createSubscribedEvent({
                      index: ++subIndex,
                      platformTime:
                          'platformTime' in envelope.payload
                              ? isISO(envelope.payload.platformTime)
                                  ? envelope.payload.platformTime
                                  : null
                              : null,
                  })
                : createUpdateEvent(getResult(envelope.payload as Exclude<P, TSubscribed>));

        return createSyncedValueDescriptor(value);
    });
}
