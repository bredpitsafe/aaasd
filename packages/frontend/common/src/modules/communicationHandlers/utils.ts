import type { TraceId } from '@common/utils';
import { generateTraceId, getNowISO } from '@common/utils';
import { invert, isObject } from 'lodash-es';
import type { Observable } from 'rxjs';
import { of, shareReplay } from 'rxjs';
import { filter, map, take } from 'rxjs/operators';

import { ModuleFactory } from '../../di';
import type { TEnvelope, TSendEnvelope, TSendPayload, WithPayload } from '../../lib/BFFSocket/def';
import type { TSocketName, TSocketStruct, TSocketURL } from '../../types/domain/sockets.ts';
import { type TSocketMap } from '../../types/domain/sockets.ts';
import { getRandomUint32 } from '../../utils/random';
import { ModuleSocketList } from '../socketList';

export function prepareEnvelopeToTracing(
    envelope: TEnvelope<WithPayload<unknown>>,
): TEnvelope<WithPayload<undefined>> {
    return {
        ...envelope,
        payload: undefined,
    };
}

export function createEnvelope<P extends TSendPayload>(
    payload: P,
    traceId?: TraceId,
    correlationId?: number,
): TSendEnvelope<P> {
    return {
        payload,
        traceId: traceId ?? generateTraceId(),
        correlationId: correlationId ?? generateCorrelationId(),
        timestamp: getNowISO(),
    };
}

export function generateCorrelationId(): number {
    return getRandomUint32();
}

export const ModuleGetSocketStruct = ModuleFactory((ctx) => {
    const { sockets$ } = ModuleSocketList(ctx);

    const mapUrlToName$ = sockets$.pipe(
        filter((sockets): sockets is TSocketMap => sockets !== undefined),
        map((sockets) => invert(sockets) as Record<TSocketURL, TSocketName>),
        shareReplay(1),
    );

    return (target: TSocketURL | TSocketStruct): Observable<TSocketStruct> => {
        return isObject(target)
            ? of(target as TSocketStruct)
            : mapUrlToName$.pipe(
                  map((mapUrlToName): TSocketStruct => {
                      return {
                          name: mapUrlToName[target as TSocketURL] ?? 'Unknown',
                          url: target,
                      };
                  }),
                  take(1),
              );
    };
});
