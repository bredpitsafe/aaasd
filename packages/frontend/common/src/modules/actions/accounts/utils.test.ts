import type { TUpdatableRealAccount } from './def.ts';
import { ACCOUNT_SECRET } from './def.ts';
import { getSendBodyAccounts } from './utils.ts';

const CREDENTIALS = {
    name: 'name',
    key: 'key',
    secret: 'secret',
    passphrase: 'passphrase',
    isInternal: false,
};

const RESET_CREDENTIALS = { ...CREDENTIALS, secret: '', passphrase: null };
const KEEP_CREDENTIALS = { ...CREDENTIALS, secret: ACCOUNT_SECRET, passphrase: ACCOUNT_SECRET };

const ACCOUNT: TUpdatableRealAccount = {
    credentials: [CREDENTIALS],
    name: 'test',
};

describe('accounts handlers utils', () => {
    describe('getSendBodyAccounts', () => {
        it('set new field value', () => {
            expect(getSendBodyAccounts([ACCOUNT])).toMatchSnapshot();
        });

        it('reset field value', () => {
            const ACCOUNT_RESET_FIELDS = {
                ...ACCOUNT,
                credentials: [RESET_CREDENTIALS],
            };

            expect(getSendBodyAccounts([ACCOUNT_RESET_FIELDS])).toMatchSnapshot();
        });

        it('keep field value', () => {
            const ACCOUNT_KEEP_FIELDS = {
                ...ACCOUNT,
                credentials: [KEEP_CREDENTIALS],
            };

            expect(getSendBodyAccounts([ACCOUNT_KEEP_FIELDS])).toMatchSnapshot();
        });
    });
});
