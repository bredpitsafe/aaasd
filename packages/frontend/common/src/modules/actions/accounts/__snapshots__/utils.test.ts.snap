// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts handlers utils getSendBodyAccounts keep field value 1`] = `
Array [
  Object {
    "credentials": Array [
      Object {
        "isInternal": false,
        "key": "key",
        "name": "name",
        "passphrase": "KeepExisting",
        "secret": "KeepExisting",
      },
    ],
    "name": "test",
  },
]
`;

exports[`accounts handlers utils getSendBodyAccounts reset field value 1`] = `
Array [
  Object {
    "credentials": Array [
      Object {
        "isInternal": false,
        "key": "key",
        "name": "name",
        "passphrase": Object {
          "Set": null,
        },
        "secret": Object {
          "Set": null,
        },
      },
    ],
    "name": "test",
  },
]
`;

exports[`accounts handlers utils getSendBodyAccounts set new field value 1`] = `
Array [
  Object {
    "credentials": Array [
      Object {
        "isInternal": false,
        "key": "key",
        "name": "name",
        "passphrase": Object {
          "Set": "passphrase",
        },
        "secret": Object {
          "Set": "secret",
        },
      },
    ],
    "name": "test",
  },
]
`;